### ISSUE TITLE
auto-encoder loss layer

### ISSUE THREAD

**Issue by daisyron** [davisking/dlib#1234]
Hi,

I tried to implement an auto-encoder network with the following general architecture:

`using cnnNetType = dlib::loss_mean_squared<dlib::upsample<2,dlib::relu<dlib::con<32,5,5,1,1,dlib::upsample<2,dlib::relu<dlib::con<32,3,3,1,1,dlib::upsample<2,dlib::relu<dlib::con<32,3,3,1,1,dlib::max_pool<2,2,2,2,dlib::relu<dlib::con<32,3,3,1,1,dlib::max_pool<2,2,2,2,dlib::relu<dlib::con<32,3,3,1,1,dlib::max_pool<2,2,2,2,dlib::relu<dlib::con<32,5,5,1,1,InputTensorType>>>>>>>>>>>>>>>>>>>;`

Due to the implementation of the `loss_mean_squared layer` it fails for compile since the training label type is float, and in auto-encoder the label type should be the same at the input tensor type.

I looked at the other loss layers in the code and non of them support, as far as I understand, the required training label type.

Is there a way in dlib to have an appropriate loss function that will enable auto-encoder training?

Thanks,

Ron.

**Comment by davisking** [comment]
loss_mean_squared is for solving regression problems.  It doesn't make sense to use as an auto encoder.  You need to use one of the losses like loss_mean_squared_multioutput, or one of the losses that have per_pixel in the name.  Or you can define your own loss function depending on what kind of autoencoder you want to train.

**Comment by daisyron** [comment]
My input image is an image with multiple channels, and an auto encoder should create the same structure at the network output. The `loss_mean_squared_multioutput` defines the `output_label_type` as `matrix<float>`. As far as I understand this will not work for me.

For auto-encoder the label type must be the same as the input type. There should be a loss function that takes the input image type as a template parameter...

**Comment by davisking** [comment]
You can always reshape your data into a `matrix<float>`. Or if that makes the code ugly or otherwise inconvenient you can define your own loss layer that does that you want.  There also already is a templated input layer that takes a range of types.
